runtime/src/accounts_db.rs

```diff
@@ -713,26 +723,35 @@ impl AccountsDB {
         // Calculate store counts as if everything was purged
         // Then purge if we can
         let mut store_counts: HashMap<AppendVecId, (usize, HashSet<Pubkey>)> = HashMap::new();
-        let storage = self.storage.read().unwrap();
-        for (key, (account_infos, _ref_count)) in &purges {
-            for (slot, account_info) in account_infos {
-                let slot_storage = storage.0.get(&slot).unwrap();
-                let store = slot_storage.get(&account_info.store_id).unwrap();
+        for (key, (account_infos, ref_count)) in purges.iter_mut() {
+            account_infos.retain(|(slot, account_info)| {
+                if old_rooted_dead_slots.contains(slot) {
+                    *ref_count = self
+                        .accounts_index
```

Comment:
removed holding storage lock above so that wouldn't dead lock with other logic that grabs `accounts_index()` lock and then the `storage lock`